resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -922,14 +897,46 @@ private Future<List<DnsRecord>> resolveAll(DnsQuestion question, DnsRecord[] add
             }
         }
 
-        // It was not A/AAAA question or there was no entry in /etc/hosts.
-        final DnsServerAddressStream nameServerAddrs =
-                dnsServerAddressStreamProvider.nameServerAddressStream(hostname);
-        new DnsRecordResolveContext(this, ch, channelReadyPromise, promise, question, additionals,
-                nameServerAddrs, maxQueriesPerResolve).resolve(promise);
+        ChannelFuture f = resolveChannelProvider.nextResolveChannel();
+        FutureListener<Object> listener = resolveChannelProvider.resolveCompletionListener(f);
+        if (listener != null) {
+            promise.addListener(listener);
+        }
+
+        if (f.isDone()) {
+            resolveAllNow(f, hostname, question, additionals, promise);
+        } else {
+            f.addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture f) {
+                    resolveAllNow(f, hostname, question, additionals, promise);
+                }
+            });
+        }
         return promise;
     }
 
+    private void resolveAllNow(ChannelFuture f, String hostname, final DnsQuestion question,
+                               final DnsRecord[] additionals, final Promise<List<DnsRecord>> promise) {
+        if (f.isSuccess()) {
+            // It was not A/AAAA question or there was no entry in /etc/hosts.
+            final DnsServerAddressStream nameServerAddrs =
+                    dnsServerAddressStreamProvider.nameServerAddressStream(hostname);
+
+            new DnsRecordResolveContext(DnsNameResolver.this, f.channel(), promise, question, additionals,
+                    nameServerAddrs, maxQueriesPerResolve).resolve(promise);
+        } else {
+            UnknownHostException e = new UnknownHostException();
```

Comment:
For all similar cases, will it provide any additional value if we add the message with some info from arguments we have in hands, like channelId, hostname, question, additionals?

Reply:
done